actors/evm/src/lib.rs

```diff
@@ -0,0 +1,190 @@
+        ActorDowncast, ActorError,
+    },
+    fvm_ipld_blockstore::Blockstore,
+    fvm_ipld_encoding::tuple::*,
+    fvm_ipld_encoding::RawBytes,
+    fvm_ipld_hamt::Hamt,
+    fvm_shared::error::*,
+    fvm_shared::{MethodNum, METHOD_CONSTRUCTOR},
+    num_derive::FromPrimitive,
+    num_traits::FromPrimitive,
+};
+
+#[cfg(feature = "fil-actor")]
+fil_actors_runtime::wasm_trampoline!(EvmContractActor);
+
+/// Maximum allowed EVM bytecode size.
+/// The contract code size limit is 24kB.
+const MAX_CODE_SIZE: usize = 24 << 10;
+
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum Method {
+    Constructor = METHOD_CONSTRUCTOR,
+    InvokeContract = 2,
+}
+
+pub struct EvmContractActor;
+impl EvmContractActor {
+    pub fn constructor<BS, RT>(rt: &mut RT, params: ConstructorParams) -> Result<(), ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        rt.validate_immediate_caller_accept_any()?;
+
+        if params.bytecode.len() > MAX_CODE_SIZE {
+            return Err(ActorError::illegal_argument(format!(
+                "EVM byte code length ({}) is exceeding the maximum allowed of {MAX_CODE_SIZE}",
+                params.bytecode.len()
+            )));
+        }
+
+        if params.bytecode.is_empty() {
+            return Err(ActorError::illegal_argument("no bytecode provided".into()));
+        }
+
+        // initialize contract state
+        let init_contract_state_cid =
+            Hamt::<_, U256, U256>::new(rt.store()).flush().map_err(|e| {
+                ActorError::unspecified(format!("failed to flush contract state: {e:?}"))
+            })?;
+        // create an instance of the platform abstraction layer -- note: do we even need this?
+        let system = System::new(rt, init_contract_state_cid).map_err(|e| {
+            ActorError::unspecified(format!("failed to create execution abstraction layer: {e:?}"))
+        })?;
+        // create a new execution context
+        let mut exec_state = ExecutionState::new(Bytes::copy_from_slice(&params.input_data));
+        // identify bytecode valid jump destinations
+        let bytecode = Bytecode::new(&params.bytecode)
+            .map_err(|e| ActorError::unspecified(format!("failed to parse bytecode: {e:?}")))?;
+        // invoke the contract constructor
+        let exec_status = execute(&bytecode, &mut exec_state, &system)
+            .map_err(|e| ActorError::unspecified(format!("EVM execution error: {e:?}")))?;
+
+        if !exec_status.reverted
+            && exec_status.status_code == StatusCode::Success
+            && !exec_status.output_data.is_empty()
+        {
+            // constructor ran to completion successfully and returned
+            // the resulting bytecode.
+            let contract_bytecode = exec_status.output_data;
+
+            let contract_state_cid = system.flush_state()?;
+
+            let state = State::new(
+                rt.store(),
+                RawBytes::new(contract_bytecode.to_vec()),
```

Comment:
nit: This is a mouthful to read as a parameter, would probably extract it to a local var and name it for what it is (evm_bytecode_bytes, or something).